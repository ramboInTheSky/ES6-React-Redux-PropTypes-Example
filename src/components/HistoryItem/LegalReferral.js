import React from 'react';
import PropTypes from 'prop-types';
import { formatting, Table } from 'nhh-styles';
import { legalReferralPropType } from '../../compositions/ActivityHistoryItem';
import { Label, Title } from './components';
import { status as legalReferralStatus } from '../../constants/legalReferral';

const LegalReferral = ({
  originalType,
  progressStatus,
  owner,
  formName,
  approvalDetail,
  lastModifiedAt,
  lastModifiedBy,
  createdAt,
  createdBy,
  status,
  referralPack,
  labels,
}) => {
  let isApprovedOrRejected;
  let isRejectedLabelApplicable;
  switch (progressStatus) {
    case legalReferralStatus.complete: {
      isApprovedOrRejected = true;
      isRejectedLabelApplicable = false;
      break;
    }
    case legalReferralStatus.rejected: {
      isApprovedOrRejected = true;
      isRejectedLabelApplicable = true;
      break;
    }
    default: {
      isApprovedOrRejected = false;
      isRejectedLabelApplicable = false;
    }
  }

  const mainTableData = [
    [
      <Label>{labels.createdAt}</Label>,
      <span data-bdd={`ActivityHistoryItem-legalReferral-date`}>
        {formatting.formatDate(createdAt, formatting.dateTimeFormat)}
      </span>,
    ],
    [
      <Label>{labels.createdBy}</Label>,
      <span data-bdd={`ActivityHistoryItem-legalReferral-rasiedBy`}>{createdBy}</span>,
    ],
    [
      <Label>{labels.formName}</Label>,
      <span data-bdd={`ActivityHistoryItem-legalReferral-with`}>{formName}</span>,
    ],
    [
      <Label>{labels.progressStatus}</Label>,
      <span data-bdd={`ActivityHistoryItem-legalReferral-with`}>{progressStatus}</span>,
    ],
    [
      <Label>{labels.owner}</Label>,
      <span data-bdd={`ActivityHistoryItem-legalReferral-with`}>{owner.name}</span>,
    ],
    [
      <Label>{labels.link}</Label>,
      <span data-bdd={`ActivityHistoryItem-legalReferral-with`}>
        {referralPack.legalReferralPack.link}
      </span>,
    ],
    [
      <Label>{labels.lastModifiedOn}</Label>,
      <span data-bdd={`ActivityHistoryItem-legalReferral-lastModified`}>
        {formatting.formatDate(lastModifiedAt, formatting.dateTimeFormat)}
        <span data-bdd={`ActivityHistoryItem-legalReferral-by`}>
          {' '}
          <Label>{labels.by}</Label> {lastModifiedBy}
        </span>
      </span>,
    ],
    [
      <Label>{labels.status}</Label>,
      <span data-bdd={`ActivityHistoryItem-legalReferral-modified`}>{status}</span>,
    ],
  ];

  const approvedRejected = [
    [
      <Label>{isRejectedLabelApplicable ? labels.rejected : labels.approval}</Label>,
      <span data-bdd={`ActivityHistoryItem-legalReferral-approval`}>
        {formatting.formatDate(approvalDetail.decisionDate, formatting.dateTimeFormat)}
        <span data-bdd={`ActivityHistoryItem-legalReferral-approval-by`}>
          {' '}
          <Label>{labels.by}</Label> {approvalDetail.userName}
        </span>
      </span>,
    ],
  ];

  const finalTable = isApprovedOrRejected ? [...mainTableData, ...approvedRejected] : mainTableData;

  return (
    <div>
      <Title>{originalType}</Title>
      <Table
        addExtraTd={false}
        data-bdd="ActivityHistoryItemTable"
        data={finalTable}
        striped
        tdWrap
        topAlign
      />
    </div>
  );
};

LegalReferral.propTypes = {
  approvalDetail: PropTypes.shape({
    decisionDate: PropTypes.string,
    userName: PropTypes.string,
  }).isRequired,
  createdAt: PropTypes.string.isRequired,
  createdBy: PropTypes.string.isRequired,
  formName: PropTypes.string.isRequired,
  labels: PropTypes.shape(legalReferralPropType).isRequired,
  lastModifiedAt: PropTypes.string.isRequired,
  lastModifiedBy: PropTypes.string.isRequired,
  originalType: PropTypes.string.isRequired,
  owner: PropTypes.shape({
    name: PropTypes.string.isRequired,
  }).isRequired,
  progressStatus: PropTypes.string.isRequired,
  referralPack: PropTypes.shape({
    legalReferralPack: PropTypes.shape({
      link: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  status: PropTypes.string.isRequired,
};

export default LegalReferral;
