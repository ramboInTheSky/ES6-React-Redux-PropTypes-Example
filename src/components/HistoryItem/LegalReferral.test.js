import React from 'react';
import { shallow } from 'enzyme';
import * as nhhStyles from 'nhh-styles';
import LegalReferral from './LegalReferral';

nhhStyles.formatting.formatDate = jest.fn(() => '10/08/2018 - 8:52pm');

describe('<LegalReferral />', () => {
  let props;

  beforeEach(() => {
    process.env.TZ = 'UTC';
    props = {
      id: 'f326549a-fb4d-43cd-855d-e5909dd951a1',
      progressStatus: 'CourtPackRequiresHomApproval',
      owner: {
        id: 'f7fe2806-6aa1-e811-80ce-005056825b41',
        name: 'CRMAppPool NonProd',
        type: 'systemuser',
      },
      formName: 'PRH Legal Referral Form | prh-legal-case-referral',
      approvalDetail: { userName: '', decisionDate: '', isApproved: false },
      lastModifiedAt: '2018-10-24T14:35:29Z',
      lastModifiedBy: 'CRMAppPool NonProd',
      createdAt: '2018-10-24T14:35:29Z',
      createdBy: 'CRMAppPool NonProd',
      status: 'In Progress',
      formServiceSubmissionId: 'ea1c9c52-d060-4cfc-82b6-c3e8740aa685',
      referralPack: {
        legalReferralForm: { link: '', type: 'letter', status: '', errorMessages: null },
        legalReferralPack: {
          link: 'this is a link',
          type: 'letter',
          status: 'Success',
          errorMessages: null,
        },
      },
      originalType: 'Legal Referral',
      labels: {
        approval: 'Approval on',
        by: 'by',
        createdAt: 'Referral date',
        createdBy: 'Raised by contact',
        formName: 'Legal case referral type',
        lastModifiedOn: 'Last modified on',
        link: 'Link to Referral',
        owner: 'Owner',
        progressStatus: 'string',
        rejected: 'Rejected on',
        status: 'Status',
      },
    };
  });

  it('renders correctly', () => {
    expect(shallow(<LegalReferral {...props} />)).toMatchSnapshot();
  });

  it('renders correctly when complete ', () => {
    expect(
      shallow(<LegalReferral {...{ ...props, progressStatus: 'Complete' }} />)
    ).toMatchSnapshot();
  });

  it('renders correctly when rejected', () => {
    expect(
      shallow(<LegalReferral {...{ ...props, progressStatus: 'HeaderDocRejected' }} />)
    ).toMatchSnapshot();
  });
});
