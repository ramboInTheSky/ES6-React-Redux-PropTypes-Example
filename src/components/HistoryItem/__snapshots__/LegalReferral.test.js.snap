// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LegalReferral /> renders correctly 1`] = `
<div>
  <Styled(styled.h3)>
    Legal Referral
  </Styled(styled.h3)>
  <Table
    addExtraTd={false}
    addExtraTh={true}
    data={
      Array [
        Array [
          <styled.span>
            Referral date
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-date"
      >
            10/08/2018 - 8:52pm
      </span>,
        ],
        Array [
          <styled.span>
            Raised by contact
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-rasiedBy"
      >
            CRMAppPool NonProd
      </span>,
        ],
        Array [
          <styled.span>
            Legal case referral type
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-with"
      >
            PRH Legal Referral Form | prh-legal-case-referral
      </span>,
        ],
        Array [
          <styled.span>
            string
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-with"
      >
            CourtPackRequiresHomApproval
      </span>,
        ],
        Array [
          <styled.span>
            Owner
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-with"
      >
            CRMAppPool NonProd
      </span>,
        ],
        Array [
          <styled.span>
            Link to Referral
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-with"
      >
            this is a link
      </span>,
        ],
        Array [
          <styled.span>
            Last modified on
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-lastModified"
      >
            10/08/2018 - 8:52pm
            <span
                  data-bdd="ActivityHistoryItem-legalReferral-by"
            >
                   
                  <styled.span>
                        by
                  </styled.span>
                   
                  CRMAppPool NonProd
            </span>
      </span>,
        ],
        Array [
          <styled.span>
            Status
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-modified"
      >
            In Progress
      </span>,
        ],
      ]
    }
    data-bdd="ActivityHistoryItemTable"
    headerData={null}
    striped={true}
    tdWrap={true}
    topAlign={true}
  />
</div>
`;

exports[`<LegalReferral /> renders correctly when complete  1`] = `
<div>
  <Styled(styled.h3)>
    Legal Referral
  </Styled(styled.h3)>
  <Table
    addExtraTd={false}
    addExtraTh={true}
    data={
      Array [
        Array [
          <styled.span>
            Referral date
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-date"
      >
            10/08/2018 - 8:52pm
      </span>,
        ],
        Array [
          <styled.span>
            Raised by contact
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-rasiedBy"
      >
            CRMAppPool NonProd
      </span>,
        ],
        Array [
          <styled.span>
            Legal case referral type
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-with"
      >
            PRH Legal Referral Form | prh-legal-case-referral
      </span>,
        ],
        Array [
          <styled.span>
            string
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-with"
      >
            Complete
      </span>,
        ],
        Array [
          <styled.span>
            Owner
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-with"
      >
            CRMAppPool NonProd
      </span>,
        ],
        Array [
          <styled.span>
            Link to Referral
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-with"
      >
            this is a link
      </span>,
        ],
        Array [
          <styled.span>
            Last modified on
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-lastModified"
      >
            10/08/2018 - 8:52pm
            <span
                  data-bdd="ActivityHistoryItem-legalReferral-by"
            >
                   
                  <styled.span>
                        by
                  </styled.span>
                   
                  CRMAppPool NonProd
            </span>
      </span>,
        ],
        Array [
          <styled.span>
            Status
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-modified"
      >
            In Progress
      </span>,
        ],
        Array [
          <styled.span>
            Approval on
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-approval"
      >
            10/08/2018 - 8:52pm
            <span
                  data-bdd="ActivityHistoryItem-legalReferral-approval-by"
            >
                   
                  <styled.span>
                        by
                  </styled.span>
                   
                  
            </span>
      </span>,
        ],
      ]
    }
    data-bdd="ActivityHistoryItemTable"
    headerData={null}
    striped={true}
    tdWrap={true}
    topAlign={true}
  />
</div>
`;

exports[`<LegalReferral /> renders correctly when rejected 1`] = `
<div>
  <Styled(styled.h3)>
    Legal Referral
  </Styled(styled.h3)>
  <Table
    addExtraTd={false}
    addExtraTh={true}
    data={
      Array [
        Array [
          <styled.span>
            Referral date
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-date"
      >
            10/08/2018 - 8:52pm
      </span>,
        ],
        Array [
          <styled.span>
            Raised by contact
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-rasiedBy"
      >
            CRMAppPool NonProd
      </span>,
        ],
        Array [
          <styled.span>
            Legal case referral type
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-with"
      >
            PRH Legal Referral Form | prh-legal-case-referral
      </span>,
        ],
        Array [
          <styled.span>
            string
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-with"
      >
            HeaderDocRejected
      </span>,
        ],
        Array [
          <styled.span>
            Owner
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-with"
      >
            CRMAppPool NonProd
      </span>,
        ],
        Array [
          <styled.span>
            Link to Referral
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-with"
      >
            this is a link
      </span>,
        ],
        Array [
          <styled.span>
            Last modified on
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-lastModified"
      >
            10/08/2018 - 8:52pm
            <span
                  data-bdd="ActivityHistoryItem-legalReferral-by"
            >
                   
                  <styled.span>
                        by
                  </styled.span>
                   
                  CRMAppPool NonProd
            </span>
      </span>,
        ],
        Array [
          <styled.span>
            Status
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-modified"
      >
            In Progress
      </span>,
        ],
        Array [
          <styled.span>
            Rejected on
      </styled.span>,
          <span
            data-bdd="ActivityHistoryItem-legalReferral-approval"
      >
            10/08/2018 - 8:52pm
            <span
                  data-bdd="ActivityHistoryItem-legalReferral-approval-by"
            >
                   
                  <styled.span>
                        by
                  </styled.span>
                   
                  
            </span>
      </span>,
        ],
      ]
    }
    data-bdd="ActivityHistoryItemTable"
    headerData={null}
    striped={true}
    tdWrap={true}
    topAlign={true}
  />
</div>
`;
